### ISSUE TITLE
display expression values in 2d bins

### ISSUE THREAD

**Issue by slowkow** [maximilianh/cellBrowser#162]
I'd like to ask if it would be possible to change the way the data is displayed in the main window.

As far as I know, there are no scRNA-seq data browsers that use 2d bins to show expression data. I would hope that all browsers will use 2d bins in the future.

This article from the documentation for the [datashader] python package does a great job showing why plotting colored dots is not optimal for large datasets.

[datashader]: https://datashader.org

https://datashader.org/user_guide/Plotting_Pitfalls.html

> Visualization is supposed to help you explore and understand your data, but if your visualizations are systematically misrepresenting your data because of overplotting, oversaturation, undersampling, undersaturation, underutilized range, and nonuniform colormapping, then you won't be able to discover the real qualities of your data and will be unable to make the right decisions.

I would consider this to be essential reading for people who try to use visualizations to inform decisions.

TL;DR

- bin the data in 2d (squares or hexagons) 
- show the summary (mean, median, percent > 0, etc.) for each bin with color
- for the color scale, use the full range of colors by showing quantiles (if lazy, log-transform)

This is bad:

![image](https://user-images.githubusercontent.com/209714/77200491-0a2a5400-6ac1-11ea-99b7-c0d50a76bcf1.png)

This is good:

![image](https://user-images.githubusercontent.com/209714/77200499-0f879e80-6ac1-11ea-8d3c-39a22e1d7814.png)

Since this feature is important to me, I will try to see what I can do. If I make any progress, I'll share it with you here in this issue. If you agree this would be a valuable improvement, please go ahead without me -- don't wait for any updates from me.

One idea is to try the [density heatmap][2] plot from vega.

[2]: https://vega.github.io/vega/examples/density-heatmaps/

![image](https://user-images.githubusercontent.com/209714/77200898-df8ccb00-6ac1-11ea-9f49-d479df803a79.png)

**Comment by slowkow** [comment]
At this point I've hacked together my own javascript that builds on top of your foundation. I'm thoroughly impressed with your code.

Here are my attempts with d3 and vega. Of course I'd be happy to share the code. Right now I'm still playing around — eventually I'll make a blog post or something.

## CellBrowser

Here's a gene in CellBrowser:

![image](https://user-images.githubusercontent.com/209714/77240017-869f5e80-6bb7-11ea-85c8-eae4640cb1e2.png)

## d3

I got a prototype working with [d3-hexbin], which shows the mean of the log2CPM expression value for cells in each bin:

[d3-hexbin]: https://observablehq.com/@d3/hexbin

![image](https://user-images.githubusercontent.com/209714/77240013-7a1b0600-6bb7-11ea-8ae5-3007ad21933b.png)

## vega

I got a rough prototype working with the vega [heatmap]. However, I don't know how to show the mean of log2CPM values. I posted a [new question on Stackoverflow][1], and I hope someone might be able to suggest a workaround.

[1]: https://stackoverflow.com/questions/60793133/2d-heatmap-of-mean-values-with-vega

So instead, this is actually showing the density of points in each square — weighted by the quantized expression values.

[heatmap]: https://vega.github.io/vega/examples/density-heatmaps/

![image](https://user-images.githubusercontent.com/209714/77240047-d9791600-6bb7-11ea-80b1-1c048ef3d30b.png)

**Comment by matthewspeir** [comment]
Hey, @slowkow!

This is really cool and it seems like it would be a really great feature. We're currently focusing on bringing in new datasets, so I'm not sure we have any time to dedicate to this. If you have an idea for how to implement this on the python backend, we can certainly draw hexagons in Javascript. Basically, if you can implement this at least partially and then need help integrating it with the rest of our code, @maximilianh said he would be happy to help. 

Thanks!

**Comment by maximilianh** [comment]
Hi @slowkow <https://github.com/slowkow>, just adding that D3 was too slow
for this, at least in my hands, when I tried to draw many circles. The
calculation of the x,y coordinates would probably too slow for Javascript
and bigger datasets, I assume? So a python implementation may have
advantages: the coords would be precalculated and then the javascript only
would have to draw the hexagons. Hope this makes sense, let me know if we
can help with something.

**Comment by slowkow** [comment]
Here's what I have been hacking on in the past few weeks. It seems to work pretty well. It is fun to build on top of the foundation that you built with the binary files and range requests 😀

Data from [Smillie et al 2019](http://doi.org/dqdf)

![ezgif com-video-to-gif (2)](https://user-images.githubusercontent.com/209714/79151556-a67e0880-7d98-11ea-93f3-baa939c6dc86.gif)

**Comment by maximilianh** [comment]
Hi Kamil, looks great. Do you have a link to this? Curious how you
implemented it...

**Comment by slowkow** [comment]
Hey Max, I apologize for the very long delay in response. Sometimes I forget to reply. Also, I was trying to decide for a long time whether I should contribute to your repo or create my own.

In the end, I made my own at https://github.com/slowkow/cellguide

I copied some of your files and then hacked new features until I had something that meets some of my needs. Of course, please feel free to copy anything — I kept the GPL-3 license. I like that I have the freedom to diverge in a different direction with my own repo.

I have more ideas for the future if you want to chat again sometime — let me know.
